compiler/src/iree/compiler/Dialect/Stream/Transforms/UnifyEncodingForGlobals.cpp

```diff
@@ -0,0 +1,330 @@
+}
+
+LogicalResult GlobalEncodingAnalyzer::collectGlobalEncodings() {
+  LDBG() << "--- collectGlobalEncodings ---";
+  // Walk all initializers to find encoding patterns:
+  //   %source = util.global.load @source_global
+  //   %encoded = stream.tensor.encode %source ...
+  //   util.global.store %encoded, @encoded_global
+  for (auto initOp : moduleOp.getOps<IREE::Util::InitializerOp>()) {
+    initOp.walk([&](IREE::Stream::TensorEncodeOp encodeOp) {
+      LDBG() << "  Found TensorEncodeOp: " << encodeOp;
+      Type resultType = encodeOp.getResultEncoding();
+      if (!hasRecognizedEncoding(resultType)) {
+        LDBG() << "    Skipping: no recognized encoding";
+        return;
+      }
+
+      auto storeOp = findStoreOp(encodeOp.getResult());
+      if (!storeOp) {
+        LDBG() << "    Skipping: no store found";
+        return;
+      }
+      StringRef encodedGlobalName = storeOp.getGlobalName();
+      auto encodedGlobalOp =
+          symbolTable.lookup<IREE::Util::GlobalOpInterface>(encodedGlobalName);
+      if (!encodedGlobalOp) {
+        LDBG() << "    Skipping: encoded global not found";
+        return;
+      }
+      if (encodedGlobalOp.isGlobalMutable()) {
+        LDBG() << "    Skipping: mutable encoded global";
+        return;
+      }
+      auto sourceGlobal = traceToSourceGlobal(encodeOp.getSource());
+      if (!sourceGlobal) {
+        LDBG() << "    Skipping: failed to trace to source global";
+        return;
+      }
+
+      LDBG() << "    Source: " << sourceGlobal.getGlobalName();
+      EncodedGlobalInfo encodedInfo;
+      encodedInfo.encodedGlobal = encodedGlobalOp;
+      encodedInfo.encodeOp = encodeOp;
+      encodedInfo.encodingAttr =
+          cast<RankedTensorType>(resultType).getEncoding();
+      if (auto affinityOp =
+              dyn_cast<IREE::Stream::AffinityOpInterface>(*encodeOp)) {
+        encodedInfo.affinityAttr = affinityOp.getAffinityAttr();
+      }
+      auto &sourceInfo = sourceGlobals[sourceGlobal.getGlobalName()];
+      if (!sourceInfo.sourceGlobal) {
+        sourceInfo.sourceGlobal = sourceGlobal;
+      }
+      sourceInfo.encodedVersions.push_back(encodedInfo);
+    });
+  }
+  return success();
+}
+
+IREE::Util::GlobalOpInterface
+GlobalEncodingAnalyzer::traceToSourceGlobal(Value value) {
+  // TODO(#22485): Extend to general subgraph canonicalization.
+  // Currently only handles direct global loads + passthrough ops.
+  SmallVector<Value> worklist;
+  DenseSet<Value> visited;
+  IREE::Util::GlobalOpInterface foundSourceGlobal;
+  worklist.push_back(value);
+  while (!worklist.empty()) {
+    Value current = worklist.pop_back_val();
+    if (!visited.insert(current).second) {
+      continue;
+    }
+    Operation *defOp = current.getDefiningOp();
+    if (!defOp) {
+      LDBG() << "      Bail: block argument";
+      return nullptr;
+    }
+    bool shouldContinue =
+        llvm::TypeSwitch<Operation *, bool>(defOp)
```

Comment:
This seems like maybe something to move to its own function to make this more readable/managable long-term.